### PR TITLE
[20240729] 정우용_8회차 문제 풀이 5주차

### PR DESCRIPTION
**Pull request by wooyong99**
- https://www.acmicpc.net/problem/25511
- https://www.acmicpc.net/problem/11725

### CODE REVIEW

**Comment by koosco** [comment]
25511
```
import sys

read = sys.stdin.readline
sys.setrecursionlimit(int(1e9))

def dfs(v: int, depth: int = 0):
    if scores[v] == k:
        global res
        res = depth
    for w in tree[v]:
        dfs(w, depth + 1)


n, k = map(int, read().split())
tree = {i: [] for i in range(n)}
res = None
for _ in range(n - 1):
    p, c = map(int, read().split())
    tree[p].append(c)
scores = list(map(int, read().split()))
dfs(0)
print(res)
```

**Comment by koosco** [comment]
11725
```
import sys

read = sys.stdin.readline
sys.setrecursionlimit(int(1e9))

def dfs(p: int, v: int = 1):
    res[v] = p
    for w in graph[v]:
        if not res[w]:
            dfs(v, w)

n = int(read())
res = [0] * (n + 1)
res[1] = -1
graph = {i: [] for i in range(1, n + 1)}
for _ in range(n - 1):
    s, e = map(int, read().split())
    graph[s].append(e)
    graph[e].append(s)
dfs(1)
print(*res[2:], sep='\n')
```

File: [8회차] 트리순회 /우용/S2_25511.java
Comment (ChangHee98) [inline [8회차] 트리순회 /우용/S2_25511.java:1]:
```
package day08;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class S2_25511 {
    public static void main(String[] args) throws IOException {
        BufferedReader br =  new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());
        /*
        * 0
        * |  \
        * 1   2
        * |\  |\
        * 3 4 5 6
        *       |
        *       7
        * */
        HashMap<Integer, ArrayList<Integer>> field = new HashMap<>();

        for(int i=0; i<n-1; i++){
            st = new StringTokenizer(br.readLine());
            int p = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());
            field.computeIfAbsent(p, key-> new ArrayList<>()).add(c);
        }
        int[] arr = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int t = 0;
        for(int i=0; i<n; i++){
            if(arr[i] == k){
                t = i;
            }
        }
        Queue<Integer> queue = new LinkedList<>();
        queue.add(0);
        if(t == 0){
            System.out.println(0);
            return;
        }
        int[] level = new int[n];
        level[0] = 0;
        while(!queue.isEmpty()){
            int current = queue.poll();
            if(field.get(current) != null){
                for(int v : field.get(current)){
                    level[v] = level[current] + 1;
                    if(v == t){
                        System.out.println(level[v]);
                        queue.clear();
                        break;
                    }
                    queue.add(v);
                }
            }
        }
    }
}

```

**Comment by MaengSung** [comment]
25511

import sys

sys.setrecursionlimit(int(1e6))
read = sys.stdin.readline

n,k = map(int,read().rstrip().split())
tree = [[] for _ in range(n)]
visited = [False] *n

for i in range(n-1):
    p,c = map(int, read().rstrip().split())
    tree[p].append(c)
    tree[c].append(p)
    
score = list(map(int,read().rstrip().split()))

def dfs(s,d):
    visited[s] = True
    if k == score[s]:
        return d
    for node in tree[s]:
        if not visited[node]:
            result = dfs(node,d+1)
            if result is not None:
                return result
    return None
print(dfs(0,0))

**Comment by MaengSung** [comment]
11725

import sys

sys.setrecursionlimit(int(1e6))
read = sys.stdin.readline

def find(graph, vertex, visit):
    for i in graph[vertex]:
        if not visit[i]:
            visit[i] = vertex
            find(graph, i, visit)

n = int(read().rstrip())

tree = [[] for _ in range(n+1)]
parent = [None for _ in range(n+1)]

for _ in range(n-1):
    s,e = map(int, read().rstrip().split())
    tree[s].append(e)
    tree[e].append(s)

find(tree, 1, parent)

for i in range(2,n+1):
    print(parent[i])

**Pull request merged**